zkvm/docs/zkvm-spec.md

```diff
@@ -1390,49 +1408,93 @@ Fails if:
+    T.commit("anchor", vm.last_anchor)
+    next_anchor = T.challenge_bytes("vm_anchor")
+    contract_anchor = T.challenge_bytes("contract_anchor")
+    ```
+4. Updates the [VM’s last anchor](#vm-state) with a ratcheted `next_anchor`:
+5. Uses ratcheted `contract_anchor` in the [output structure](#output-structure).
+6. Adds an [output entry](#output-entry) to the [transaction log](#transaction-log).
 
 Immediate data `k` is encoded as [LE32](#le32).
 
+Fails if VM’s [last anchor](#vm-state) is missing.
+
 
 #### contract
 
 _items... pred_ **contract:_k_** → _contract_
 
 1. Pops [predicate](#predicate) `pred` from the stack.
 2. Pops `k` items from the stack.
-3. Creates a contract with the `k` items as a payload and the predicate.
+3. Creates a contract with the `k` items as a payload, the predicate `pred`, and anchor set to the [VM’s last anchor](#vm-state).
 4. Pushes the contract onto the stack.
+5. Update the [VM’s last anchor](#vm-state) with the [contract ID](#contract-id) of the new contract.
 
 Immediate data `k` is encoded as [LE32](#le32).
 
+Fails if VM’s [last anchor](#vm-state) is missing.
 
 #### nonce
 
-_predicate_ **nonce** → _contract_
+_pred blockid_ **nonce** → _contract_
+
+1. Pops [32-byte string](#data-type) `blockid` from the stack.
+2. Pops [predicate](#predicate) `pred` from the stack.
+3. Computes the nonce [anchor](#anchor) value:
+    ```
+    T = Transcript("ZkVM.nonce")
+    T.commit("blockid", blockid)
+    T.commit("predicate", pred)
+    T.commit("maxtime", LE64(tx.maxtime))
+    nonce_anchor = T.challenge_bytes()
+    ```
+4. Pushes a new [contract](#contract-type) with:
+    * an empty [payload](#contract-payload)
+    * predicate set to `pred`
+    * anchor set to `nonce_anchor`.
+5. Sets the [VM’s last anchor](#vm-state) to [contract ID](#contract-id) of that contract.
+6. Adds [nonce entry](#nonce-entry) to the [transaction log](#transaction-log) with the `blockid`, transaction [maxtime](#time-bounds) and `nonce_anchor`.
+
+Fails if:
+* `blockid` is a not a 32-byte [data string](#data-type).
+* `pred` is not a valid [point](#point).
+
+**Discussion**
+
+A nonce serves two purposes:
 
-1. Pops [predicate](#predicate) from the stack.
-2. Pushes a new [contract](#contract-type) with an empty [payload](#contract-payload) and this predicate to the stack.
-3. Adds [nonce entry](#nonce-entry) to the [transaction log](#transaction-log) with the predicate and transaction [maxtime](#time-bounds).
-4. Sets the [VM uniqueness flag](#vm-state) to `true`.
+1. It provides a unique anchor for an issuing transaction if other anchors are not available (i.e., from other values already on the stack);
+2. It binds the entire transaction to a particular blockchain, protecting not only against cross-blockchain replays, but also potential blockchain forks due to compromise of the old block-signing keys. This is a necessary feature for stateless signing devices that rely on blockchain proofs.
```

Comment:
> block-signing keys

Not clear to me that we're going to have those. Perhaps only in the consensus layer, not in the blockchain data model per se.

Reply:
yeah, that paragraph is very much out of scope. How about: 
```suggestion
2. It binds the entire transaction to a particular blockchain, protecting not only against cross-blockchain replays, but also potential blockchain forks. This is a necessary feature for stateless signing devices that rely on blockchain proofs.
```

Reply:
That looks good, but note that block signatures are also mentioned in one or two other places.

Reply:
oops